source/Attribute.cpp

```diff
@@ -0,0 +1,425 @@
+	if(category == SHIELD_GENERATION && effectType == HULL)
+		effectName = "repair";
+
+	string composite;
+	// shortcuts: "energy generation energy" is "energy generation", etc.
+	if(static_cast<int>(category) == effectType && category <= CLOAKING)
+		composite = categoryName;
+	else if(category == PROTECTION || category == RESISTANCE || (category == DAMAGE && effect != FORCE))
+		composite = effectName + ((effectName == "" || categoryName == "") ? "" : " ") + categoryName;
+	else
+		composite = categoryName + ((effectName == "" || categoryName == "") ? "" : " ") + effectName;
+
+	if(IsRelative())
+		composite = "relative " + composite;
+	if(IsMultiplier())
+		composite += " multiplier";
+	return composite;
+}
+
+
+
+namespace {
+	// Cached mappings between the two formats. The indexing of newToOld is offset by one compared to the enum values.
+	string newToOld[ATTRIBUTE_CATEGORY_COUNT + 1][ATTRIBUTE_EFFECT_COUNT * 4 + 1][ATTRIBUTE_EFFECT_COUNT * 4 + 1];
+	unordered_map<string, Attribute*> oldToNew;
+}
+
+
+
+// Maps the attribute to the legacy single string format.
+string Attribute::GetLegacyName() const
+{
+	if(!cached)
+		PrepareCache();
+	return newToOld[category + 1][effect + 1][secondary + 1];
+}
+
+
+
+// Gets the attribute for the specified token, if any.
+Attribute *Attribute::Parse(const string &token)
+{
+	if(!cached)
+		PrepareCache();
+	auto it = oldToNew.find(token);
+	if(it == oldToNew.end())
+		return nullptr;
+	return it->second;
+}
+
+
+
+// Creates a cache for faster data loading.
+void Attribute::PrepareCache()
+{
+	for(int effect = -1; effect < ATTRIBUTE_EFFECT_COUNT * 4; effect++)
+	{
+		for(int category = -1; category < ATTRIBUTE_CATEGORY_COUNT; category++)
+		{
+			for(int secondary = -1; secondary < (ATTRIBUTE_EFFECT_COUNT * 4) * (category == RESISTANCE); secondary++)
+			{
+				Attribute *a = new Attribute(static_cast<AttributeCategory>(category), static_cast<AttributeEffect>(effect),
+						static_cast<AttributeEffect>(secondary));
+				string s = a->CalculateLegacyName();
+				newToOld[category + 1][effect + 1][secondary + 1] = s;
+				if(oldToNew.find(s) == oldToNew.end())
+					oldToNew[s] = a;
+				else
+					delete a;
+			}
+		}
+	}
+	// ensure that effect names are always recognized
+	for(int effect = 0; effect < ATTRIBUTE_EFFECT_COUNT * 4; effect++)
+	{
+		string s = GetEffectName(static_cast<AttributeEffect>(effect));
+		if(oldToNew.find(s) == oldToNew.end())
+			oldToNew[s] = new Attribute(static_cast<AttributeCategory>(-1), static_cast<AttributeEffect>(effect),
+					static_cast<AttributeEffect>(-1));;
+	}
+	cached = true;
+}
+
+
+
+// Checks whether the specified combination is fully supported by the engine.
+bool Attribute::IsSupported() const
+{
+	if(effect == -1)
+		return false;
+	if(secondary != -1)
+	{
+		if(category != RESISTANCE)
+			return false;
+		if(effect >= ATTRIBUTE_EFFECT_COUNT || effect == PIERCING)
+			return false;
+		if(!Attribute(category, effect).IsSupported())
+			return false;
```

Comment:
I'd put them all into the same if with () and stuff but mb it's more readable this way idk

Reply:
Idk if it's more readable, but it shows the logic better and is easier to tweak.

Reply:
yeah you can keep it like this if you want